ert_shared/cli/main.py

```diff
@@ -84,3 +84,6 @@ def run_cli(args):
     if model.hasRunFailed():
         clear_global_state()
         raise ErtCliError  # the monitor has already reported the error message
+
+    if ee_config is not None:
+        ee_config.cleanup()
```

Comment:
Seems this would not run if ERT fails, i.e. if there are failing realizations—a common occurrence. 

I was thinking, wouldn't an `ExitStack` be good here? You can then push onto that stack `out.close()`, `clear_global_state()` as well as the  `ee_config.cleanup()`.

https://docs.python.org/3/library/contextlib.html#contextlib.ExitStack

Might be too bleeding edge for 3.6 though.

Reply:
Good catch...

I actually consider dropping the explicit cleanup. From the [Python docs](https://docs.python.org/3.8/library/socket.html)

> Sockets are automatically closed when they are garbage-collected, but it is recommended to close() them explicitly, or to use a with statement around them.

and since the only purpose of EvaluatorServerConfig.cleanup() is to close the socket holding the the port (not used actively), I'm tempted to delegate that to GC. What sayest thou?

Reply:
GC is preemptive so I'd go for the recommended approach. We have a massive refactor of the ensemble evaluator coming up where ownership of the socket is likely to shift, so explicit > implicit?

Reply:
Unless @sondre disagrees with that latter part…

Reply:
The socket being closed is the one ensuring nobody else grabs the port. Sockets provided to clients are duplicates of said port and are assumed to be managed by whoever requests them.

In other words: The socket in question is private to and intimately coupled with the EvaluatorServerConfig-instance. Thus, IMO, they should live and die together. In fact, accidentily closing this socket while keeping the EvaluatorServerConfig-instance alive is likely to cause trouble. I.e. if we keep cleanup() we might need to add an init()-method to restore the EvaluatorServerConfig-instance to working state.

Reply:
Just pushed a version without cleanup() for review. Adding cleanup() can be done in separate commit...

Reply:
I think it would be best to do it explicitly yes. 

> if we keep cleanup() we might need to add an init()-method to restore the EvaluatorServerConfig-instance to working state.

Is there any scenario where we would need to do this?

Reply:
> > if we keep cleanup() we might need to add an init()-method to restore the EvaluatorServerConfig-instance to working state.
> 
> Is there any scenario where we would need to do this?

Not really AFAICS.

The argument is that calling cleanup (aka closing   __socket:handle_) makes the EvaluatorServerConfig-instance unusable. So IMO it is safer drop cleanup and rather close  ___socket_handle_ when the EvaluatorServerConfig-instance is not referenced anymore, i.e. during GC.

This way we make it impossible to request a socket from an invalid EvaluatorServerConfig-instance. An EvaluatorServerConfig-instance is always able to return a valid socket.

Reply:
Alright, makes sense. Let's take the pragmatic approach this time and let GC handle the socket closing.